sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebSocketClient/WebSocketClient.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Messaging.WebPubSub.Clients
+{
+    internal sealed partial class WebSocketClient : IWebSocketClient
+    {
+        private readonly ClientWebSocket _socket;
+        private readonly Uri _uri;
+        private readonly string _protocol;
+        private readonly MemoryBufferWriter _buffer;
+
+        private readonly SemaphoreSlim _sendLock = new SemaphoreSlim(1);
+
+        public WebSocketClient(Uri uri, string protocol)
+        {
+            _protocol = protocol;
+            _socket = new ClientWebSocket();
+            _socket.Options.AddSubProtocol(_protocol);
+            _uri = uri;
+            _buffer = new MemoryBufferWriter();
+        }
+
+        public void Dispose()
+        {
+            _sendLock.Dispose();
+            _socket.Dispose();
+            _buffer.Dispose();
+        }
+
+        public async Task ConnectAsync(CancellationToken token)
+        {
+            WebPubSubClientEventSource.Log.WebSocketConnecting(_protocol);
+
+            await _socket.ConnectAsync(_uri, token).ConfigureAwait(false);
+        }
+
+        public async Task SendAsync(ReadOnlyMemory<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
+        {
+            await _sendLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                await _socket.SendAsync(buffer, messageType, endOfMessage, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                _sendLock.Release();
+            }
+        }
+
+        public async Task<WebSocketReadResult> ReceiveOneFrameAsync(CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            if (_socket.State == WebSocketState.Closed)
+            {
+                return new WebSocketReadResult(default, true);
+            }
+
+            _buffer.Reset();
```

Comment:
interesting.. do you want it to be thread safe

Reply:
It's thread safe. buffer uses `ThreadStatic`

Reply:
This doesn't have to be thread safe as no concurrent invocation